examples/reward_LM_gsm8k/inference.py

```diff
@@ -0,0 +1,119 @@
+            eos_token_id = [364,402,512,756]
+        elif self.base_model.model.config.architectures[0] == 'Qwen2ForCausalLM':
+            eos_token_id = [198,271,382,624,151645]
+        else:
+            assert isinstance(self.base_model.model, transformers.LlamaForCausalLM)###need to be modified for other model
+            eos_token_id = [13]
+        for i in range((self.n_sc - 1) // self.bs + 1):
+            local_bs = min(self.bs, self.n_sc - i * self.bs)
+            outputs += self.base_model.generate([inputs] * local_bs,
+                                            hide_input=True,
+                                            do_sample=do_sample,
+                                            top_p=0.95,
+                                            temperature=self.temperature,
+                                            eos_token_id=eos_token_id).text
+        return [o.strip() for o in outputs], question
+
+def main(base_lm:Literal['hf', 'google', 'openai', 'anthropic','exllama'],model_dir, lora_dir=None, mem_map=None, batch_size=1, prompt="examples/cot_gsm8k/prompts/cot.json", resume=0, log_dir=None, temperature=0, n_sc=1, quantized='int8', quantized_reward = None, reward_lm = None, reward_dir = None):
+
+    if base_lm == "openai":
+        base_model = OpenAIModel("gpt-4-1106-preview", additional_prompt="ANSWER")
+    elif base_lm == "google":
+        base_model = BardCompletionModel("gemini-pro", additional_prompt="ANSWER")
+    elif base_lm == "anthropic":
+        base_model = ClaudeModel("claude-3-opus-20240229", additional_prompt="ANSWER")
+    elif base_lm == "hf":
+        base_model = HFModel(model_dir, model_dir, quantized=quantized)
+    elif base_lm == "exllama":
+        base_model = ExLlamaModel(model_dir, max_batch_size = n_sc, max_seq_length=5000, max_new_tokens=1000, lora_dir=None)
+
+    reward_tokenizer = None
+    assert reward_dir is not None, "Reward model path cannot be none for CoT + Reward"
+    if reward_lm == "openai":
+        raise NotImplementedError("GPTCompletionModel does not support reward model output")
+    elif reward_lm == "google":
+        raise NotImplementedError("BardCompletionModel does not support reward model output")
+    elif reward_lm == "anthropic":
+        raise NotImplementedError("ClaudeModel does not support reward model output")
+    elif reward_lm == "hf":
+        reward_model = HFModelReward(model_pth = reward_dir, tokenizer_pth = reward_dir,quantized=quantized_reward, device_map = "cuda:0")
+    elif reward_lm == "exllama":
+        reward_model = ExLlamaModel(reward_dir, max_batch_size = n_sc, max_seq_length=5000, lora_dir=None)
+    
+    
+    
+    with open(prompt) as f:
+        prompt = json.load(f)
+
+    reasoner = CoT_Reward_Reasoner(base_model, temperature=temperature, n_sc=n_sc, bs=batch_size, main_prompt = prompt)
+    utils_reward = Reward_CoT_Utils(reward_model = reward_model)
```

Comment:
As we discussed last time, we want to implement a `RandomShooting` algorithm. Now the reward model becomes an evaluation part, which is conceptually misleading.

Reply:
I have fixed this. There is no longer a `Reward_CoT_Utils` anymore. It is closer to what a planning algorithm would look like. I generate a bunch of candidate CoT chains, I score each of them with reward model & then pass the best one to the evaluator (which has not changes & is exactly same as the `cot/gsm8k` example)